<a href="https://colab.research.google.com/github/megajoules8/data_science_worksheets/blob/main/hospital_data_project1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
## Assignment 5 - Pandas

In [ ]:
## Required Assignment 5.2: Case Study - Patient Arrivals in Singapore’s Major Public Hospitals

#### Learning Outcomes Addressed
Analyse the patient arrival dataset of Singapore's major public hospitals using Pandas

#### Assignment Instructions
In this assignment, you will analyse the patient arrival dataset of Singapore's major public hospitals.

In [ ]:
import pandas as pd
import os
cwd = os.getcwd()
df = pd.read_csv('EDdata_new.csv')

In [ ]:
df

Case Hospital_Name  REGIS_DATE REGIS_TIME  reg_sec Triage Time  \
0       92408          KTPH    7/4/2012    9:48:33    35313     9:58:12   
1       54452           KKH   7/10/2011   16:21:05    58865    16:23:42   
2       28303           CGH   3/10/2011    3:57:45    14265     4:00:00   
3      121169           SGH  16/10/2011    4:08:47    14927     4:10:00   
4      146488          TTSH  24/10/2011    3:09:47    11387     3:14:21   
...       ...           ...         ...        ...      ...         ...   
41827   38592           CGH  29/10/2011   14:07:15    50835    14:17:00   
41828   20523            AH   2/10/2011    7:18:06    26286     7:18:00   
41829   30317           CGH   8/10/2011   12:59:42    46782    12:59:00   
41830  126705           SGH  31/10/2011    9:55:18    35718    10:06:00   
41831  110904           NUH   11/4/2012   11:20:27    40827    12:11:00   

       triage_sec Triage_Class    Age Gender     Race  
0           35892            P2  40.0      M  Chinese  
1           59022            P2   0.0      M  Chinese  
2           14400            P3  33.0      M   Indian  
3           15000            P3  53.0      F    Malay  
4           11661            P3  23.0      M   Others  
...           ...           ...   ...    ...      ...  
41827       51420            P2  82.0      F  Chinese  
41828       26280            P1  88.0      F  Chinese  
41829       46740            P2  59.0      M    Malay  
41830       36360            P3  54.0      F  Chinese  
41831       43860            P3  19.0      F  Chinese  

[41832 rows x 11 columns]

In [ ]:
## Case Questions

### QUESTION 1
Do male Singaporeans have preferences over different hospitals to attend in case of an emergency? Please remember to delete the patient visits to (KKH in the dataset KKH is a Women's and Children's hospital.) If an emergency happens, the male patient will basically not be sent to KKH.

In [ ]:
criter1 = df.Gender == "M"
criter2 = df.Hospital_Name != "KKH"
print(df[criter1 & criter2])

         Case Hospital_Name  REGIS_DATE REGIS_TIME  reg_sec Triage Time  \
0       92408          KTPH    7/4/2012    9:48:33    35313     9:58:12   
2       28303           CGH   3/10/2011    3:57:45    14265     4:00:00   
4      146488          TTSH  24/10/2011    3:09:47    11387     3:14:21   
5       93761          KTPH   11/4/2012    0:13:43      823     2:51:08   
6       95762          KTPH  16/04/2012   23:14:41    83681     1:05:13   
...       ...           ...         ...        ...      ...         ...   
41819   41738           CGH    6/4/2012   14:41:57    52917    14:45:00   
41820   94560          KTPH  13/04/2012   15:32:04    55924    16:07:46   
41823  152502          TTSH    9/4/2012    8:43:28    31408     8:47:51   
41824  100722           NUH   2/10/2011   14:49:36    53376    15:18:00   
41829   30317           CGH   8/10/2011   12:59:42    46782    12:59:00   

       triage_sec Triage_Class    Age Gender     Race  
0           35892            P2  40.0      

In [ ]:
fil_1 = criter1&criter2
df_male = df.loc[fil_1].copy()
print(df_male)

         Case Hospital_Name  REGIS_DATE REGIS_TIME  reg_sec Triage Time  \
0       92408          KTPH    7/4/2012    9:48:33    35313     9:58:12   
2       28303           CGH   3/10/2011    3:57:45    14265     4:00:00   
4      146488          TTSH  24/10/2011    3:09:47    11387     3:14:21   
5       93761          KTPH   11/4/2012    0:13:43      823     2:51:08   
6       95762          KTPH  16/04/2012   23:14:41    83681     1:05:13   
...       ...           ...         ...        ...      ...         ...   
41819   41738           CGH    6/4/2012   14:41:57    52917    14:45:00   
41820   94560          KTPH  13/04/2012   15:32:04    55924    16:07:46   
41823  152502          TTSH    9/4/2012    8:43:28    31408     8:47:51   
41824  100722           NUH   2/10/2011   14:49:36    53376    15:18:00   
41829   30317           CGH   8/10/2011   12:59:42    46782    12:59:00   

       triage_sec Triage_Class    Age Gender     Race  
0           35892            P2  40.0      

In [ ]:
hosp_name = df_male['Hospital_Name'].unique()
print(hosp_name)

['KTPH' 'CGH' 'TTSH' 'AH' 'SGH' 'NUH']


In [ ]:
out_table = dict.fromkeys(hosp_name,0)
print(out_table)

{'KTPH': 0, 'CGH': 0, 'TTSH': 0, 'AH': 0, 'SGH': 0, 'NUH': 0}


In [ ]:
for h in hosp_name:
    filter_hos = df.Hospital_Name == h
    out_table[h]=df.loc[filter_hos].sum()
result=pd.Series(out_table)
print(result)

KTPH    Case                                          ...
CGH     Case                                          ...
TTSH    Case                                          ...
AH      Case                                          ...
SGH     Case                                          ...
NUH     Case                                          ...
dtype: object


In [ ]:
male_patients = df.Gender == 'M'
print(male_patients)

0         True
1         True
2         True
3        False
4         True
         ...  
41827    False
41828    False
41829     True
41830    False
41831    False
Name: Gender, Length: 41832, dtype: bool


In [ ]:
grpy = df.groupby('Hospital_Name')
print(grpy)
for h, group in grpy:
    print(h)
    print()

AH

CGH

KKH

KTPH

NUH

SGH

TTSH



In [ ]:
cols_need = ['Hospital_Name','Gender']
df_short = df[cols_need]
print(df_short)

      Hospital_Name Gender
0              KTPH      M
1               KKH      M
2               CGH      M
3               SGH      F
4              TTSH      M
...             ...    ...
41827           CGH      F
41828            AH      F
41829           CGH      M
41830           SGH      F
41831           NUH      F

[41832 rows x 2 columns]


In [ ]:
## Code -> Q1
Gender_filter = df_short['Gender'] == 'M'
hosp_filter = df_short['Hospital_Name'] != 'KKH'
df_short_males = df_short.loc[Gender_filter, :]
df_short_hosp = df_short_males.loc[hosp_filter, :]

print(df_short_hosp)
hosp_name = df_male['Hospital_Name'].unique()

grpy = df_short_hosp.groupby('Hospital_Name').count()
print(grpy)

      Hospital_Name Gender
0              KTPH      M
2               CGH      M
4              TTSH      M
5              KTPH      M
6              KTPH      M
...             ...    ...
41819           CGH      M
41820          KTPH      M
41823          TTSH      M
41824           NUH      M
41829           CGH      M

[20330 rows x 2 columns]
               Gender
Hospital_Name        
AH               1433
CGH              4301
KTPH             3886
NUH              2904
SGH              3435
TTSH             4371


##### RESPONSE TO QUESTION 1 (based on result shown in the above df)
Yes. Male patients have preferences over different hospitals. The Hospital data detailing "Patient Arrivals in Singapore's Major Public Hospitals" show the visits made by Male patients ranging from 1433 to 4371. This concludes that TTSH recorded the most number of male patients @4371, making TTSH the most popular hopital visted by male patients.

### QUESTION 2
Are the patients’ waiting time distributions similar across different public hospitals?

In [ ]:
## Question 2

df['wait_time'] = 0
n_pat = len(df)
df_reg = df['reg_sec']
df_tri = df['triage_sec']
df_wait = df['wait_time']

df_wait = 24*60*60 - df_reg + df_tri
filt_1 = df['triage_sec'] > df['reg_sec']
df_wait [filt_1] = df_tri [filt_1] - df_reg [filt_1]

df ['wait_time'] = df_wait
df['wait_min'] = df_wait/60
df.head()

Case Hospital_Name  REGIS_DATE REGIS_TIME  reg_sec Triage Time  \
0   92408          KTPH    7/4/2012    9:48:33    35313     9:58:12   
1   54452           KKH   7/10/2011   16:21:05    58865    16:23:42   
2   28303           CGH   3/10/2011    3:57:45    14265     4:00:00   
3  121169           SGH  16/10/2011    4:08:47    14927     4:10:00   
4  146488          TTSH  24/10/2011    3:09:47    11387     3:14:21   

   triage_sec Triage_Class    Age Gender     Race  wait_time  wait_min  
0       35892            P2  40.0      M  Chinese        579  9.650000  
1       59022            P2   0.0      M  Chinese        157  2.616667  
2       14400            P3  33.0      M   Indian        135  2.250000  
3       15000            P3  53.0      F    Malay         73  1.216667  
4       11661            P3  23.0      M   Others        274  4.566667

##### RESPONSE TO QUESTION 2 (based on result shown in the above df)
No. The patients waiting time range 3:09:47 to 16:21:05, which concludes TTSH in the least in terms of wait time.

### QUESTION 3
To make a staffing plan, which decides the number of nurses and doctors needed to serve patients, a deep understanding of patient arrivals is crucial. The staffing plan in practice will be made on an hourly basis (24 intervals) every day. Thus, please create a new column, REGIS_HOUR in df. Moreover, the patients’ arrival pattern may vary by the day of a month. Please also create a new column, REGIS_DAY in df.

In [ ]:
df['REGIS_HOUR'] = pd.to_datetime(df.REGIS_DATE + ' ' + df.REGIS_TIME).dt.hour
df['REGIS_DAY'] = pd.to_datetime(df.REGIS_DATE + ' ' + df.REGIS_TIME).dt.day
df['regis'] = pd.to_datetime(df.REGIS_DATE + ' ' + df.REGIS_TIME)
df

Case Hospital_Name  REGIS_DATE REGIS_TIME  reg_sec Triage Time  \
0       92408          KTPH    7/4/2012    9:48:33    35313     9:58:12   
1       54452           KKH   7/10/2011   16:21:05    58865    16:23:42   
2       28303           CGH   3/10/2011    3:57:45    14265     4:00:00   
3      121169           SGH  16/10/2011    4:08:47    14927     4:10:00   
4      146488          TTSH  24/10/2011    3:09:47    11387     3:14:21   
...       ...           ...         ...        ...      ...         ...   
41827   38592           CGH  29/10/2011   14:07:15    50835    14:17:00   
41828   20523            AH   2/10/2011    7:18:06    26286     7:18:00   
41829   30317           CGH   8/10/2011   12:59:42    46782    12:59:00   
41830  126705           SGH  31/10/2011    9:55:18    35718    10:06:00   
41831  110904           NUH   11/4/2012   11:20:27    40827    12:11:00   

       triage_sec Triage_Class    Age Gender     Race  wait_time     wait_min  \
0           35892            P2  40.0      M  Chinese        579     9.650000   
1           59022            P2   0.0      M  Chinese        157     2.616667   
2           14400            P3  33.0      M   Indian        135     2.250000   
3           15000            P3  53.0      F    Malay         73     1.216667   
4           11661            P3  23.0      M   Others        274     4.566667   
...           ...           ...   ...    ...      ...        ...          ...   
41827       51420            P2  82.0      F  Chinese        585     9.750000   
41828       26280            P1  88.0      F  Chinese      86394  1439.900000   
41829       46740            P2  59.0      M    Malay      86358  1439.300000   
41830       36360            P3  54.0      F  Chinese        642    10.700000   
41831       43860            P3  19.0      F  Chinese       3033    50.550000   

       REGIS_HOUR  REGIS_DAY               regis  
0               9          4 2012-07-04 09:48:33  
1              16         10 2011-07-10 16:21:05  
2               3         10 2011-03-10 03:57:45  
3               4         16 2011-10-16 04:08:47  
4               3         24 2011-10-24 03:09:47  
...           ...        ...                 ...  
41827          14         29 2011-10-29 14:07:15  
41828           7         10 2011-02-10 07:18:06  
41829          12         10 2011-08-10 12:59:42  
41830           9         31 2011-10-31 09:55:18  
41831          11          4 2012-11-04 11:20:27  

[41832 rows x 16 columns]

##### RESPONSE TO QUESTION 3 (based on result shown in the above df)
Two new columns, "REGIS_HOUR" and "REGIS_DAY" created in df as shown above.

### QUESTION 4
Find out the average number of patient arrivals in each hour of a day. To answer this question, we assume the arrival pattern is similar across different days and only utilise the data in 2011.

In [ ]:
average_patients_per_hour = pd.read_csv('EDdata_new.csv', parse_dates=[[2, 3]])
#pd.read_csv?

In [ ]:
average_patients_per_hour['REGIS_DATE'] = average_patients_per_hour.REGIS_DATE_REGIS_TIME.dt.date
average_patients_per_hour['REGIS_HOUR'] = average_patients_per_hour.REGIS_DATE_REGIS_TIME.dt.hour

In [ ]:
average_patients_per_hour[['Case', 'REGIS_DATE', 'REGIS_HOUR', 'REGIS_DATE_REGIS_TIME']][average_patients_per_hour.REGIS_DATE_REGIS_TIME <'2012-01-01'].groupby(['REGIS_HOUR'])['REGIS_HOUR'].count()

REGIS_HOUR
0      652
1      474
2      404
3      345
4      289
5      284
6      338
7      507
8      764
9     1107
10    1365
11    1414
12    1211
13    1225
14    1253
15    1116
16    1156
17     949
18     932
19    1074
20    1268
21    1176
22     994
23     849
Name: REGIS_HOUR, dtype: int64

In [ ]:
average_patients_per_hour[['Case', 'REGIS_HOUR', 'REGIS_DATE_REGIS_TIME']][average_patients_per_hour.REGIS_DATE_REGIS_TIME <'2012-01-01'].groupby(['REGIS_HOUR'])['REGIS_HOUR'].count() / average_patients_per_hour[average_patients_per_hour.REGIS_DATE_REGIS_TIME < '2012-01-01'].REGIS_DATE.nunique()

REGIS_HOUR
0     21.032258
1     15.290323
2     13.032258
3     11.129032
4      9.322581
5      9.161290
6     10.903226
7     16.354839
8     24.645161
9     35.709677
10    44.032258
11    45.612903
12    39.064516
13    39.516129
14    40.419355
15    36.000000
16    37.290323
17    30.612903
18    30.064516
19    34.645161
20    40.903226
21    37.935484
22    32.064516
23    27.387097
Name: REGIS_HOUR, dtype: float64

##### RESPONSE TO QUESTION 4 (based on result shown in the above df)
I use datetime to parse the information from the dataframe to simplify the calculations. I used data < '2012-01-01' to utilize the data in 2011. My answer for question four reitertaes the answer for question three. For simplicity, i just defined a new DataFrame, created the necessary columns again and then used the .groupby function to group the necessary columns, the .count function to get the amount needed to calculate the mean and the next line of code performs the calculation and prints out the average patients per hour for year 2011.

## Question 5
The assumption that the arrival pattern is similar across different days is too strong to be true. Let's discuss the weekday effect (including Saturday and Sunday) on the arrival pattern of patients. Please create a WEEKDAY column in df. For example, if a patient's arrival occurred on 01/10/2011, the corresponding value in WEEKDAYcolumn is Saturday.

In [ ]:
average_patients_per_hour['REGIS_Weekday'] = average_patients_per_hour.REGIS_DATE_REGIS_TIME.dt.weekday

In [ ]:
average_patients_per_hour

REGIS_DATE_REGIS_TIME    Case Hospital_Name  reg_sec Triage Time  \
0       2012-07-04 09:48:33   92408          KTPH    35313     9:58:12   
1       2011-07-10 16:21:05   54452           KKH    58865    16:23:42   
2       2011-03-10 03:57:45   28303           CGH    14265     4:00:00   
3       2011-10-16 04:08:47  121169           SGH    14927     4:10:00   
4       2011-10-24 03:09:47  146488          TTSH    11387     3:14:21   
...                     ...     ...           ...      ...         ...   
41827   2011-10-29 14:07:15   38592           CGH    50835    14:17:00   
41828   2011-02-10 07:18:06   20523            AH    26286     7:18:00   
41829   2011-08-10 12:59:42   30317           CGH    46782    12:59:00   
41830   2011-10-31 09:55:18  126705           SGH    35718    10:06:00   
41831   2012-11-04 11:20:27  110904           NUH    40827    12:11:00   

       triage_sec Triage_Class    Age Gender     Race  REGIS_DATE  REGIS_HOUR  \
0           35892            P2  40.0      M  Chinese  2012-07-04           9   
1           59022            P2   0.0      M  Chinese  2011-07-10          16   
2           14400            P3  33.0      M   Indian  2011-03-10           3   
3           15000            P3  53.0      F    Malay  2011-10-16           4   
4           11661            P3  23.0      M   Others  2011-10-24           3   
...           ...           ...   ...    ...      ...         ...         ...   
41827       51420            P2  82.0      F  Chinese  2011-10-29          14   
41828       26280            P1  88.0      F  Chinese  2011-02-10           7   
41829       46740            P2  59.0      M    Malay  2011-08-10          12   
41830       36360            P3  54.0      F  Chinese  2011-10-31           9   
41831       43860            P3  19.0      F  Chinese  2012-11-04          11   

       REGIS_Weekday  
0                  2  
1                  6  
2                  3  
3                  6  
4                  0  
...              ...  
41827              5  
41828              3  
41829              2  
41830              0  
41831              6  

[41832 rows x 13 columns]

##### RESPONSE TO QUESTION 5 (based on result shown in the above df)
"REGIS_Weekday" created in the last column in df.

### QUESTION 6
Using the WEEKDAY column, please find out the average number of patient arrivals in each hour by weekday categories. Your answer should be a 7-by-24 table.

In [ ]:
pivoted = (average_patients_per_hour.query('REGIS_DATE_REGIS_TIME < "2012-01-01"')
           .pivot_table(
               'Case',
               index='REGIS_HOUR',
               columns='REGIS_Weekday',
               aggfunc='count')
           )
pivoted

REGIS_Weekday    0    1    2    3    4    5    6
REGIS_HOUR                                      
0              117   64   53  137   86  101   94
1               79   49   42   81   52   84   87
2               71   45   36   85   47   62   58
3               58   21   44   74   40   53   55
4               51   21   25   57   34   45   56
5               51   23   21   63   23   44   59
6               63   34   32   73   40   48   48
7              100   45   43  103   71   70   75
8              144   78   71  160  102  108  101
9              209  106  106  224  133  172  157
10             282  128  119  252  167  216  201
11             295  132  140  261  168  193  225
12             207  114  104  241  155  163  227
13             213  113  133  246  113  213  194
14             215  104  123  233  153  200  225
15             196   87   94  220  141  199  179
16             210  104  121  230  157  158  176
17             157   85   96  196  117  147  151
18             148   77  101  177  107  146  176
19             187   96  107  215  117  178  174
20             231  104  120  238  163  197  215
21             192  106  118  213  146  169  232
22             171   87  101  185  126  164  160
23             138   92   75  154  103  128  159